src/transformers/generation/utils.py

```diff
@@ -2318,7 +2318,7 @@ def greedy_search(
             )
 
         # keep track of which sequences are already finished
-        unfinished_sequences = torch.ones(input_ids.shape[0], dtype=torch.long, device=input_ids.device)
```

Comment:
WDYT about these changes to handle 3d logit vectors and input ids? (the extra dim is over codebooks) If we're ok with this I can propagate the changes to the other generate methods

Reply:
Typically, we have logits with dimensionality: 
```
(bsz, seq_len, vocab_size)
```

Here, we have an extra dimension over codebooks:
```
(bsz, num_codebooks, seq_len, vocab_size)
```

This is leading to the logit scores breaking as they pass through some of our logits processors - what do we think about reshaping the logits to:
```
(bsz * num_codebooks, seq_len, vocab_size)
```
As we do for beam search?

The only difficulty I see is that we do our argmax, we want to do it over the vocab dim:
```python
torch.argmax(logit_scores, -1)
```
In a way that we predict the argmax token for **each** codebook

Reply:
I think that it would be great if we don't have to do it in this PR, but it might be a good time to update this to support more feature sizes

Reply:
I don't think we should have any `if model_type == "musicgen"` in the `generate` method tbh

Reply:
@patrickvonplaten the alternative is custom generate code, which is much more painful to maintain in the long run. 

I'm also planning to work on refactoring generate to be a composable sequence of functions, in which case musicgen could simply wrap the token selection step to reshape the scores.

Reply:
@patrickvonplaten the alternative is custom generate code, which is much more painful to maintain in the long run (and will lose on new features).

I'm also planning to work on refactoring generate to be a composable sequence of functions, in which case musicgen could simply wrap the token selection step to reshape the scores. Keeping the custom code in generate makes the transition simpler, IMO, as the divergent path is immediately clear.